### PR TITLE
[9.0][ADD] Migrated business_requirement_deliverable_report module from v8 to v9

### PR DESCRIPTION
**Pull request by sudhir-serpentcs**
<s> WIP until #25  and #26  is merged</s>

Migrated business_requirement_deliverable_report module from v8 to v9.
@dreispt @pedrobaeza @jbeficent @moylop260 Please review.

### CODE REVIEW

File: business_requirement_deliverable_report/README.rst
Comment (elicoidal) [inline business_requirement_deliverable_report/README.rst]: 9.0

**Comment by elicoidal** [comment]
:+1:  (waiting for #25 to be merged)

**Review by elicoidal — CHANGES_REQUESTED** [review]

File: business_requirement_deliverable_report/__openerp__.py
Comment (elicoidal) [inline business_requirement_deliverable_report/__openerp__.py]: add the icon

**Comment by sudhir-serpentcs** [comment]
Rebased the PR.
@dreispt @pedrobaeza @jbeficent @moylop260 @elicoidal

**Review by elicoidal — APPROVED** [review]

**Comment by elicoidal** [comment]
@sudhir-serpentcs tested functionally. 3 small details for improvements in the 3 reports:

* when saving the name should be "BR000001-test.pdf" (name+"-"+description)
* there is a big gap between the header and the start of the page ("Business Requirement Document" text). Not sure this is not due to standard Odoo header spacing. Please confirm whether you can improve or not (simple spacing is enough)
* The spacing between "Business Requirement Document" and "BR000001-test" is too big. "BR000001-test" should be centered below "Business Requirement Document"
(check example in Runbot)

**Comment by sudhir-serpentcs** [comment]
@elicoidal Thanks for you feedback. I will do the needful.
I guess there is not functionality of having dynamic report name (based on the value of the fields).

>when saving the name should be "BR000001-test.pdf" (name+"-"+description)

Do you mean by report name or attachment name of the report?

**Comment by elicoidal** [comment]
@sudhir-serpentcs I mean attachment name of the report, thanks for pointing out

**Comment by sudhir-serpentcs** [comment]
@elicoidal Added file attachment option and its name for the generated reports.

File: business_requirement_deliverable_report/report/report.xml
Comment (elicoidal) [inline business_requirement_deliverable_report/report/report.xml]: following the logic below: "'BR- '+((object.name+object.description)+'.pdf')"

**Review by elicoidal — CHANGES_REQUESTED** [review]
Not sure what was done but runbot says it is not working ;).
My downloaded attachments have respectively the names:
* "Business Requirement Report.pdf",
* "Business Requirement Deliverable Report.pdf" and
* "Business Requirement Deliverable Resource Report.pdf"

Did I miss something?

Besides following issues are not yet attended (you plan them for later?):
> * there is a big gap between the header and the start of the page ("Business Requirement Document" text). Not sure this is not due to standard Odoo header spacing. Please confirm whether you can improve or not (simple spacing is enough)
> * The spacing between "Business Requirement Document" and "BR000001-test" is too big. "BR000001-test" should be centered below "Business Requirement Document"
(check example in Runbot)

**Comment by elicoidal** [comment]
@sudhir-serpentcs Any news on this front?

File: business_requirement_deliverable_report/report/report.xml
Comment (sudhir-serpentcs) [reply business_requirement_deliverable_report/report/report.xml]: @elicoidal Done

**Comment by sudhir-serpentcs** [comment]
>- there is a big gap between the header and the start of the page ("Business Requirement Document" text). Not sure this is not due to standard Odoo header spacing. Please confirm whether you can improve or not (simple spacing is enough)
>- The spacing between "Business Requirement Document" and "BR000001-test" is too big. "BR000001-test" should be centered below "Business Requirement Document"
(check example in Runbot)

@elicoidal Done. Please review.

**Comment by elicoidal** [comment]
@sudhir-serpentcs thanks for the update. 

> there is a big gap between the header and the start of the page ("Business Requirement Document" text). Not sure this is not due to standard Odoo header spacing. Please confirm whether you can improve or not (simple spacing is enough)

OK!

> The spacing between "Business Requirement Document" and "BR000001-test" is too big. "BR000001-test" should be centered below "Business Requirement Document"
(check example in Runbot)

You can reduce the space between "Business Requirement Document" and "BR000001-test" to half the current size.

> My downloaded attachments have respectively the names:
    "Business Requirement Report.pdf",
    "Business Requirement Deliverable Report.pdf" and
    "Business Requirement Deliverable Resource Report.pdf"

The naming of the files is still incorrect when you save them.

**Comment by elicoidal** [comment]
@sudhir-serpentcs any feedback here?

**Comment by sudhir-serpentcs** [comment]
My downloaded attachments have respectively the names:
    "Business Requirement Report.pdf",
    "Business Requirement Deliverable Report.pdf" and
    "Business Requirement Deliverable Resource Report.pdf"
     The naming of the files is still incorrect when you save them.
@elicoidal Do you mean to say when we download the report, its name should be "BR000001-test.pdf" (name+"-"+description) like this?

If yes, I am not sure how can we give dynamic name to the report.

**Comment by elicoidal** [comment]
> Do you mean to say when we download the report, its name should be "BR000001-test.pdf" (name+"-"+description) like this?

Yes this is what I mean. Maybe you can have a check at how the SO or invoice reports  are added to the mails. I think the names are specific when attached.
@sudhir-serpentcs 

If no simple solution then forget about the requirement.

**Comment by sudhir-serpentcs** [comment]
@elicoidal Yes, you are right. In Sales or Invoice, the report name is specified in the Email Template.
I did some research and found that we can do it by overriding the report / controller methods.
I am leaving this for now and give update on following point:

     You can reduce the space between "Business Requirement Document" and "BR000001-test" to half the current size.

**Comment by sudhir-serpentcs** [comment]
You can reduce the space between "Business Requirement Document" and "BR000001-test" to half the current size.
@elicoidal I think it will be bit hard to remove the gap as system adds BR tag in between as the report template is called. Please check the screenshot.

br_report

**Comment by elicoidal** [comment]
@sudhir-serpentcs Thanks for the feedback and clear explanation.
Forget this requirement.

**Comment by elicoidal** [comment]
AFAIU we have only left the name of the report. right @sudhir-serpentcs ?

**Comment by sudhir-serpentcs** [comment]
AFAIU we have only left the name of the report. right @sudhir-serpentcs ?
@elicoidal Thanks Eric. Please check my [previous comment](https://github.com/OCA/business-requirement/pull/31#issuecomment-284954820) for having report name.

**Comment by elicoidal** [comment]
@sudhir-serpentcs I saw your comment and agree. go ahead

**Comment by sudhir-serpentcs** [comment]
@elicoidal Added code (controller) for the dynamic report name but you know we don't know how to make UT for controllers and that has reduced the codecov.
dyn_report_name

**Comment by elicoidal** [comment]
@sudhir-serpentcs I think we can live with the decrease coverage for the moment.

**Review by elicoidal — APPROVED** [review]

**Comment by elicoidal** [comment]
To be ported to v8 and v10

**Comment by dreispt** [comment]
Do we really need to extend the Report generation controller?
That sound like bad practice, and adds tight coupling between two components that should be independent.

Do we really need to customize the output file name?
I feel that that is a different feature that should be implemented in an OCA/report module, that could then be leveraged here.

**Comment by elicoidal** [comment]
@dreispt 
>  That sound like bad practice, and adds tight coupling between two components that should be independent.

Not sure why this is bad practice. What I know is that when you have a project with dozens of BR, you want to have all of them properly identified and I currently spend manual time just to adapt that.
For me it is more the other way round: not having the filename properly generated at printing that looks like bad practice.
In short: if there are good reasons not to change the controller I will follow them. If there is a better way to centralize this feature then let's go for it: when it is ready when could remove the controller in this module.
What do you think?

**Comment by dreispt** [comment]
@elicoidal The feature is good. 
Having it implemented together in this module looks clumsy.
Tomorrow someone will have a similar need and will also overwrite the report controller when building some XYZ repo.

For me these are two distinct features.
The controller code should be isolated in it's own module.
This module here should only care to provide the report itself. 
I believe that having both these modules installed will have the desired effect, even if they don't depend on each other.

**Comment by elicoidal** [comment]
@sudhir-serpentcs 
Can you split the controller part in a separate module business_requirement_deliverable_report_filenaming?
Once done we can merge this one.

**Comment by dreispt** [comment]
@sudhir-serpentcs @elicoidal AFAICS the report file naming feature is universal and not specific to BRs. I think it should be proposed to https://github.com/OCA/reporting-engine.

**Comment by elicoidal** [comment]
> @sudhir-serpentcs @elicoidal AFAICS the report file naming feature is universal and not specific to BRs. I think it should be proposed to https://github.com/OCA/reporting-engine.

Fine by me: I have no clue about naming though (base_filename?)

**Comment by elicoidal** [comment]
@sudhir-serpentcs when refactored (removed the controller part), please port this module to v8 and v10

**Comment by sudhir-serpentcs** [comment]
@elicoidal Ok I will separate the report file naming feature (controller) in a new module (report_base_filename). Should I propose the PR in https://github.com/OCA/reporting-engine repo?
We need dynamic reporting name module in v8-v9 only because v10 already have this feature in base report module.

**Comment by elicoidal** [comment]
@sudhir-serpentcs go ahead

**Comment by sudhir-serpentcs** [comment]
@elicoidal Removed the code to have dynamic report name. I will add the reference of the PR that will have dynamic report name feature.

**Comment by serpentcs-dev1** [comment]
@sudhir-serpentcs @elicoidal 

Please find the PR here https://github.com/OCA/reporting-engine/pull/114 which will help to have the 
dynamic feature name on report.

**Comment by elicoidal** [comment]
@dreispt can we move forward and merge this PR?

**Review by dreispt — APPROVED** [review]
Thanks for your patience!

**Comment by dreispt** [comment]
PS: Reading the discussion at  OCA/reporting-engine#114, I'm now sure that doing the separation was the right thing to do.

**Comment by elicoidal** [comment]
@dreispt indeed thanks for your help :)

**Pull request merged**
